04-ruby-mvc-sinatra/databases/sql_basics.md

```diff
@@ -0,0 +1,333 @@
+
+Here's a quick demo... (probably don't need to follow along).
+
+```sql
+help -- general help
+\?   -- help with psql commands
+\h   -- help with SQL commands
+\l   -- Lists all databases
+
+CREATE DATABASE wdi7;
+# What changed?
+\l
+
+-- What happens if we don't use a semicolon?
+
+\c wdi7 -- Connect to pbj database
+
+\d -- Lists all tables
+
+CREATE TABLE students (
+  id SERIAL PRIMARY KEY,
+  first_name VARCHAR NOT NULL,
+  last_name VARCHAR NOT NULL,
+  quote TEXT, birthday VARCHAR,
+  ssn INT NOT NULL UNIQUE
+);
+
+\d
+
+SELECT * FROM students;
+
+INSERT INTO students (first_name, last_name) VALUES ('Robin', 'Thomas');
+-- This won't work!
+
+INSERT INTO students (first_name, last_name, quote, birthday, ssn) VALUES ('Robin', 'Thomas', 'Two goldfish are in a tank. One says, "Know how to drive this thing?"', 'April 1', 8675309);
+SELECT * FROM students;
+
+UPDATE students SET first_name = 'Robert' WHERE first_name = 'Robin';
+SELECT * FROM students;
+
+DELETE FROM students WHERE first_name = 'Robin';
+DELETE FROM students WHERE first_name = 'Robert';
+
+SELECT * FROM students;
+
+DROP TABLE students;
+
+DROP DATABASE wdi7;
+
+\q --quits
+```
+
+In short:
+* Backslash commands (e.g. `\l` ) are commands to navigate psql.
+* Everything else is SQL, to get / modify info in the current database.
+
+### SQL Syntax
+
+- All statements end in a semicolon
+- Whitespace doesn't matter
+- Uppercasing!
+- [Ye olde style guide](http://leshazlewood.com/software-engineering/sql-style-guide/)
+
+### Breakin' stuff
+
+If you forget a closing bracket or semicolon, `\q` will get you out.
+
+```
+SELECT *
+
+SELECT (
+
+\q
+```
+
+## Schema
+
+Every database (for a given program we're writing) will have one or more
+**tables**. Each table stores similar data, e.g. a `songs` table, or an
+`artists` table.
+
+Each table has a **schema** which defines what columns it has, including:
+
+* the column's name
+* the column's data type
+* any constraints (or rules) for that column
+
+### Common Data Types
+
+Here are some common data types for SQL DBs:
+
+- Boolean
+- Integer
+- Float
+- Text / VARCHAR
+  - VARCHAR is short, TEXT is long
+  - VARCHAR is "CHARACTER VARYING"
+- NULL
+- Date
+- Time
+- [And many more](http://www.postgresql.org/docs/9.3/interactive/datatype.html)
+
+### Constraints
+
+Constraints act as limits on the data that can go in a column.
+
+- i.e. NOT NULL and UNIQUE
+  - [And many more](http://www.postgresql.org/docs/8.1/static/ddl-constraints.html)
+- Why are constraints and data types different?
+  - You'd have to have INTEGERNOTNULL and INTEGER
+  - Like flags for CLI commands
+
+### Defining a Schema
+
+We need to define a schema for our database. It can change later on if we need
+to add / remove tables or columns, but we'll start with something simple.
+
+Instead of typing this into psql, it's easier to save it into a file, and 'run'
+that file.
+
+We'll be using this as an example today:
```

Comment:
mention something about the file ending in `.sql` and not `.rb` - very common error